src/miral/input_configuration.cpp

```diff
@@ -0,0 +1,323 @@
+/*
+ * Copyright © Canonical Ltd.
+ *
+ * This program is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 or 3,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "miral/input_configuration.h"
+
+#include "input_config.h"
+
+#include <mir/server.h>
+#include <mir/input/input_device_hub.h>
+
+#include <algorithm>
+
+class miral::InputConfiguration::Mouse::Self : public MouseInputConfiguration
+{
+    friend class miral::InputConfiguration::Self;
+    using MouseInputConfiguration::operator=;
+};
+
+class miral::InputConfiguration::Touchpad::Self : public TouchpadInputConfiguration
+{
+    friend class miral::InputConfiguration::Self;
+    using TouchpadInputConfiguration::operator=;
+};
+
+class miral::InputConfiguration::Self
+{
+public:
+    std::shared_ptr<mir::input::InputDeviceHub> input_device_hub{};
+    std::shared_ptr<InputDeviceConfig> input_device_config;
+
+    auto mouse() -> Mouse;
+    void mouse(Mouse const& val);
+    auto touchpad() -> Touchpad;
+    void touchpad(Touchpad const& val);
+
+    void apply(Mouse const& m)
+    {
+        if (input_device_hub)
+        {
+            input_device_hub->for_each_mutable_input_device([&m](auto& input_device)
+            {
+                m.self->apply_to(input_device);
```

Comment:
Question: is there a world where we have multiple mice and we want a different configuration for each? It seems unlikely

Reply:
Agreed